### ISSUE TITLE
Change packet types

### ISSUE THREAD

**Issue by udippant** [quicwg/base-drafts#1655]
For initial and 0-rtt packets, we're planning on using the 5-tuple in our L4 load balancer, and the connid for other packet types. To do this we need to look at the packet types as to whether or not we should look at the connid. However with the current packet format, we need an extra shift instruction to process whether or not we need to look at the connid. A simple reorganization of the packet types as:

Initial 0x7F
0-RTT Protected 0x7E
Retry 0x7D
Handshake 0x7C

will help us determine whether or not to use connid with one instruction : ( pkt_type ^ FF ) > 1

In environments, such as bpf program, number of instructions matter. Further, this provides an easy win, and intuitive separation of packet types with and without server chosen connection id.

**Comment by martinthomson** [comment]
I was considering a similar change, for completely different reasons.  Let's keep this as a reminder.

**Comment by udippant** [comment]
Thank you @martinthomson . Happy to submit a P.R. if you'd prefer so.

**Comment by martinthomson** [comment]
The reason this is up for review is the discussion about the first octet.  The working group mailing list has [a discussion on that](https://mailarchive.ietf.org/arch/msg/quic/N2lbdSFFxkhAYtwCCBlNHBZeCEU).  I'd prefer to wait for that.

**Comment by mikkelfj** [comment]
`((int8_t)pkt_type + (int8_t)0x40) < 0`

The add operation should overflow and set the sign if the two upper bits are set and not in any other case. clang x64 does it slightly different using a cmp 0x40 followed by a conditional set in the example below, but that could be a conditional jump. 

```
#include <stdint.h>
#include <stdio.h>

int test(uint8_t pkt_type)
{
    int8_t x = (int8_t)pkt_type;
    return (x + (int8_t)0x40) < 0;
}
```
clang -c -S -O3 test.c

```
    .section    __TEXT,__text,regular,pure_instructions
    .macosx_version_min 10, 13
    .globl  _test                   ## -- Begin function test
    .p2align    4, 0x90
_test:                                  ## @test
    .cfi_startproc
## BB#0:
    pushq   %rbp
Lcfi0:
    .cfi_def_cfa_offset 16
Lcfi1:
    .cfi_offset %rbp, -16
    movq    %rsp, %rbp
Lcfi2:
    .cfi_def_cfa_register %rbp
    xorl    %eax, %eax
    cmpb    $-64, %dil
    setl    %al
    popq    %rbp
    retq
    .cfi_endproc
                                        ## -- End function

.subsections_via_symbols
```

**Comment by mikkelfj** [comment]
This above was tesing for 0b11xxxxxx
Testing for 0xb10xxxxxx
requires one extra and instruction following clang

```
#include <stdint.h>

uint8_t test(uint8_t pkt_type)
{
    return (pkt_type & 0xc0) != 0x80;
}
```

```

    .section    __TEXT,__text,regular,pure_instructions
    .macosx_version_min 10, 13
    .globl  _test                   ## -- Begin function test
    .p2align    4, 0x90
_test:                                  ## @test
    .cfi_startproc
## BB#0:
    pushq   %rbp
Lcfi0:
    .cfi_def_cfa_offset 16
Lcfi1:
    .cfi_offset %rbp, -16
    movq    %rsp, %rbp
Lcfi2:
    .cfi_def_cfa_register %rbp
    andb    $-64, %dil
    xorl    %eax, %eax
    cmpb    $-128, %dil
    setne   %al
    popq    %rbp
    retq
    .cfi_endproc
                                        ## -- End function

.subsections_via_symbols
```

**Comment by udippant** [comment]
(pkt_type & 0xc0 != 0x80) should be fine. But curious about your type grouping that results in this (with first MSB for long header, you'd still just be using the second MSB here).

**Comment by martinthomson** [comment]
I'll go through this tomorrow, but the short answer is that we need to balance this with the need to multiplex with RTP.  (The final equation we settled on was `t & 0xa0 != 0x80`, the c was an error.)

**Comment by mjoras** [comment]
Random note: compiler explorer is very useful for comparing things like this. E.g. https://godbolt.org/z/YEOwVM

**Comment by martinthomson** [comment]
Closed by #2006.
